### ISSUE TITLE
Forbidden to mutate SDK outside of the `SdkModificator`

### ISSUE THREAD

**Issue by gcb** [KronicDeth/intellij-elixir#3655]
**Describe the bug**
Cannot select detected SDK from new project dropdown, nor select the directory.

throws:
> SEVERE - #c.i.o.p.i.ProjectJdkImpl - Forbidden to mutate SDK outside of the `SdkModificator`. Please, use `com.intellij.openapi.projectRoots.Sdk.getSdkModificator`
> SEVERE - #c.i.o.p.i.ProjectJdkImpl - Plugin to blame: Elixir version: 19.0.

**To Reproduce**
1. Install clean idea community edition
2. on WelcomeScreen, install plugin from marketplace.
3. optionally restart idea.
4. again on welcomeScreen, create new Elixir project.
5. select the detected elixir SDK from drop down.

**Expected behavior**
New project

**Screenshots**
n/a

**Logs**
```
2024-10-07 17:45:54,619 [   7085] SEVERE - #c.i.o.p.i.ProjectJdkImpl - Forbidden to mutate SDK outside of the `SdkModificator`. Please, use `com.intellij.openapi.projectRoots.Sdk.getSdkModificator`
java.lang.Throwable: Forbidden to mutate SDK outside of the `SdkModificator`. Please, use `com.intellij.openapi.projectRoots.Sdk.getSdkModificator`
        at com.intellij.openapi.diagnostic.Logger.error(Logger.java:376)
        at com.intellij.openapi.projectRoots.impl.ProjectJdkImpl.setHomePath(ProjectJdkImpl.java:122)
        at org.elixir_lang.sdk.elixir.Type$Companion.createDefaultErlangSdk(Type.kt:474)
        at org.elixir_lang.sdk.elixir.Type$Companion.createDefaultErlangSdk(Type.kt:495)
        at org.elixir_lang.sdk.elixir.Type$Companion.defaultErlangSdk(Type.kt:513)
        at org.elixir_lang.sdk.elixir.Type$Companion.configureInternalErlangSdk(Type.kt:341)
        at org.elixir_lang.sdk.elixir.Type$Companion.configureSdkPaths(Type.kt:333)
        at org.elixir_lang.sdk.elixir.Type$Companion.access$configureSdkPaths(Type.kt:251)
        at org.elixir_lang.sdk.elixir.Type.setupSdkPaths(Type.kt:188)
        at com.intellij.openapi.projectRoots.SdkType.setupSdkPaths(SdkType.java:125)
        at com.intellij.openapi.roots.ui.configuration.projectRoot.ProjectSdksModel.lambda$setupSdk$10(ProjectSdksModel.java:533)
        at com.intellij.openapi.progress.impl.CoreProgressManager$2.run(CoreProgressManager.java:328)
        at com.intellij.openapi.progress.impl.CoreProgressManager.startTask(CoreProgressManager.java:477)
        at com.intellij.openapi.progress.impl.ProgressManagerImpl.startTask(ProgressManagerImpl.java:133)
        at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressSynchronously$9(CoreProgressManager.java:567)
        at com.intellij.openapi.progress.impl.ProgressRunner.lambda$new$0(ProgressRunner.java:85)
        at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:249)
        at com.intellij.openapi.progress.ProgressManager.lambda$runProcess$0(ProgressManager.java:98)
        at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$1(CoreProgressManager.java:221)
        at com.intellij.platform.diagnostic.telemetry.helpers.TraceKt.use(trace.kt:49)
        at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:220)
        at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:660)
        at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:735)
        at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:691)
        at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:659)
        at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:79)
        at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:202)
        at com.intellij.openapi.progress.ProgressManager.runProcess(ProgressManager.java:98)
        at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$5(ProgressRunner.java:249)
        at com.intellij.openapi.progress.impl.ProgressRunner$ProgressRunnable.run(ProgressRunner.java:501)
        at com.intellij.openapi.progress.impl.ProgressRunner.lambda$launchTask$18(ProgressRunner.java:466)
        at com.intellij.util.concurrency.ChildContext$runAsCoroutine$1.invoke(propagation.kt:92)
        at com.intellij.util.concurrency.ChildContext$runAsCoroutine$1.invoke(propagation.kt:92)
        at com.intellij.util.concurrency.ChildContext.runAsCoroutine(propagation.kt:97)
        at com.intellij.util.concurrency.ChildContext.runAsCoroutine(propagation.kt:92)
        at com.intellij.openapi.progress.impl.ProgressRunner.lambda$launchTask$19(ProgressRunner.java:464)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:735)
        at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:732)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
        at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:732)
        at java.base/java.lang.Thread.run(Thread.java:1583)
2024-10-07 17:45:54,628 [   7094] SEVERE - #c.i.o.p.i.ProjectJdkImpl - IntelliJ IDEA 2024.2.3  Build #IC-242.23339.11
2024-10-07 17:45:54,628 [   7094] SEVERE - #c.i.o.p.i.ProjectJdkImpl - JDK: 21.0.4; VM: OpenJDK 64-Bit Server VM; Vendor: Arch Linux
2024-10-07 17:45:54,628 [   7094] SEVERE - #c.i.o.p.i.ProjectJdkImpl - OS: Linux
2024-10-07 17:45:54,629 [   7095] SEVERE - #c.i.o.p.i.ProjectJdkImpl - Plugin to blame: Elixir version: 19.0.1
2024-10-07 17:45:54,629 [   7095] SEVERE - #c.i.o.p.i.ProjectJdkImpl - Last Action: WelcomeScreen.CreateNewProject
2024-10-07 17:45:54,659 [   7125] SEVERE - #c.i.o.p.Task - current thread is not owner
java.lang.IllegalMonitorStateException: current thread is not owner
        at java.base/java.lang.Object.notify(Native Method)
        at org.elixir_lang.jps.HomePath.eachEbinPath(HomePath.java:50)
        at org.elixir_lang.sdk.Type.addCodePaths(Type.kt:14)
        at org.elixir_lang.sdk.erlang.Type.setupSdkPaths(Type.kt:119)
        at org.elixir_lang.sdk.elixir.Type$Companion.createDefaultErlangSdk(Type.kt:475)
        at org.elixir_lang.sdk.elixir.Type$Companion.createDefaultErlangSdk(Type.kt:495)
        at org.elixir_lang.sdk.elixir.Type$Companion.defaultErlangSdk(Type.kt:513)
        at org.elixir_lang.sdk.elixir.Type$Companion.configureInternalErlangSdk(Type.kt:341)
        at org.elixir_lang.sdk.elixir.Type$Companion.configureSdkPaths(Type.kt:333)
        at org.elixir_lang.sdk.elixir.Type$Companion.access$configureSdkPaths(Type.kt:251)
        at org.elixir_lang.sdk.elixir.Type.setupSdkPaths(Type.kt:188)
        at com.intellij.openapi.projectRoots.SdkType.setupSdkPaths(SdkType.java:125)
        at com.intellij.openapi.roots.ui.configuration.projectRoot.ProjectSdksModel.lambda$setupSdk$10(ProjectSdksModel.java:533)
        at com.intellij.openapi.progress.impl.CoreProgressManager$2.run(CoreProgressManager.java:328)
        at com.intellij.openapi.progress.impl.CoreProgressManager.startTask(CoreProgressManager.java:477)
        at com.intellij.openapi.progress.impl.ProgressManagerImpl.startTask(ProgressManagerImpl.java:133)
        at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressSynchronously$9(CoreProgressManager.java:567)
        at com.intellij.openapi.progress.impl.ProgressRunner.lambda$new$0(ProgressRunner.java:85)
        at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:249)
        at com.intellij.openapi.progress.ProgressManager.lambda$runProcess$0(ProgressManager.java:98)
        at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$1(CoreProgressManager.java:221)
        at com.intellij.platform.diagnostic.telemetry.helpers.TraceKt.use(trace.kt:49)
        at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:220)
        at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:660)
        at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:735)
        at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:691)
        at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:659)
        at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:79)
        at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:202)
        at com.intellij.openapi.progress.ProgressManager.runProcess(ProgressManager.java:98)
        at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$5(ProgressRunner.java:249)
        at com.intellij.openapi.progress.impl.ProgressRunner$ProgressRunnable.run(ProgressRunner.java:501)
        at com.intellij.openapi.progress.impl.ProgressRunner.lambda$launchTask$18(ProgressRunner.java:466)
        at com.intellij.util.concurrency.ChildContext$runAsCoroutine$1.invoke(propagation.kt:92)
        at com.intellij.util.concurrency.ChildContext$runAsCoroutine$1.invoke(propagation.kt:92)
        at com.intellij.util.concurrency.ChildContext.runAsCoroutine(propagation.kt:97)
        at com.intellij.util.concurrency.ChildContext.runAsCoroutine(propagation.kt:92)
        at com.intellij.openapi.progress.impl.ProgressRunner.lambda$launchTask$19(ProgressRunner.java:464)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:735)
        at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:732)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
        at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:732)
        at java.base/java.lang.Thread.run(Thread.java:1583)
2024-10-07 17:45:54,660 [   7126] SEVERE - #c.i.o.p.Task - IntelliJ IDEA 2024.2.3  Build #IC-242.23339.11
2024-10-07 17:45:54,660 [   7126] SEVERE - #c.i.o.p.Task - JDK: 21.0.4; VM: OpenJDK 64-Bit Server VM; Vendor: Arch Linux
2024-10-07 17:45:54,660 [   7126] SEVERE - #c.i.o.p.Task - OS: Linux
2024-10-07 17:45:54,660 [   7126] SEVERE - #c.i.o.p.Task - Plugin to blame: Elixir version: 19.0.1
2024-10-07 17:45:54,661 [   7127] SEVERE - #c.i.o.p.Task - Last Action: WelcomeScreen.CreateNewProject
```
**Desktop:**
Operating System: Arch Linux 
Kernel Version: 6.11.1-arch1-1 (64-bit)

**Erlang:**
- Installer: distro
- Version: erlang-nox 27.1-1

**Elixir:**
- Installer: distro
- Version: elixir 1.17.3-1

**Plugin:**
- Version: 19.0.1

**Additional context**
n/a

**Comment by joshuataylor** [comment]
Which IDE?

**Comment by gcb** [comment]
My bad. I think i forgot to paste this block:

```
IntelliJ IDEA 2024.2.3 (Community Edition)
Build #IC-242.23339.11, built on September 28, 2024
Runtime version: 21.0.4+7 amd64
VM: OpenJDK 64-Bit Server VM by Arch Linux
Toolkit: sun.awt.X11.XToolkit
Linux 6.11.2-arch1-1
GC: G1 Young Generation, G1 Concurrent GC, G1 Old Generation
Memory: 3926M
Cores: 12
Registry:
  ide.experimental.ui=true
  i18n.locale=
Non-Bundled Plugins:
  Dart (242.22855.32)
  org.elixir_lang (19.0.1)
  IdeaVIM (2.16.0)
  org.jetbrains.android (242.23339.11)
  com.android.tools.design (242.23339.11)
  io.flutter (81.1.3)
Kotlin: 242.23339.11-IJ
Current Desktop: KDE
```

**Comment by joshuataylor** [comment]
Thanks! I'll see what's going on.

**Comment by Fgabz** [comment]
Hello 👋 ,

any update, or any workaround ?

**Comment by joshuataylor** [comment]
I've been trying to find a better workaround for this, but it's gonna be a
bit of a hack until we can implement the new way that IntelliJ wants custom
SDKs.

I'll try and get something out soon to hitfix this and the other SDK issues
.

**Comment by aikomastboom** [comment]
sorry for not making this into a PR, I just got it to work on my machine:
```diff
Index: src/org/elixir_lang/sdk/elixir/Type.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/org/elixir_lang/sdk/elixir/Type.kt b/src/org/elixir_lang/sdk/elixir/Type.kt
--- a/src/org/elixir_lang/sdk/elixir/Type.kt	(revision b794170d35ae3ede74ee1645531b1199cad2075e)
+++ b/src/org/elixir_lang/sdk/elixir/Type.kt	(date 1754488421160)
@@ -331,7 +331,9 @@
             addDocumentationPaths(sdkModificator)
             addSourcePaths(sdkModificator)
             configureInternalErlangSdk(sdk, sdkModificator)
-            ApplicationManager.getApplication().runWriteAction { sdkModificator.commitChanges() }
+            ApplicationManager.getApplication().invokeLaterOnWriteThread ({
+                ApplicationManager.getApplication().runWriteAction { sdkModificator.commitChanges() }
+            }, ModalityState.nonModal())
         }
 
         private fun configureInternalErlangSdk(
@@ -471,15 +473,20 @@
         ): Sdk? {
             val sdkName = erlangSdkType.suggestSdkName("Default " + erlangSdkType.name, homePath)
             val projectJdkImpl = ProjectJdkImpl(sdkName, erlangSdkType)
-            projectJdkImpl.homePath = homePath
-            erlangSdkType.setupSdkPaths(projectJdkImpl)
+            var modificator = projectJdkImpl.sdkModificator
+            modificator.homePath = homePath
 
             return if (projectJdkImpl.versionString != null) {
-                ApplicationManager.getApplication().invokeAndWait(
+                ApplicationManager.getApplication().invokeLaterOnWriteThread (
                     {
-                        ApplicationManager.getApplication().runWriteAction { projectJdkTable.addJdk(projectJdkImpl) }
+                        ApplicationManager.getApplication().runWriteAction {
+                            modificator.commitChanges() // save
+                            erlangSdkType.setupSdkPaths(projectJdkImpl)
+                            projectJdkTable.addJdk(projectJdkImpl)
+                        }
                     },
-                    ModalityState.NON_MODAL,
+                    ModalityState.nonModal(),
                 )
                 projectJdkImpl
             } else {
Index: src/org/elixir_lang/sdk/erlang/Type.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/org/elixir_lang/sdk/erlang/Type.kt b/src/org/elixir_lang/sdk/erlang/Type.kt
--- a/src/org/elixir_lang/sdk/erlang/Type.kt	(revision b794170d35ae3ede74ee1645531b1199cad2075e)
+++ b/src/org/elixir_lang/sdk/erlang/Type.kt	(date 1754488421176)
@@ -2,6 +2,7 @@
 
 import com.intellij.execution.ExecutionException
 import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.application.ModalityState
 import com.intellij.openapi.diagnostic.Logger
 import com.intellij.openapi.projectRoots.Sdk
 import com.intellij.openapi.projectRoots.SdkModel
@@ -117,7 +118,9 @@
         val sdkModificator = sdk.sdkModificator
         org.elixir_lang.sdk.Type
             .addCodePaths(sdkModificator)
-        ApplicationManager.getApplication().runWriteAction { sdkModificator.commitChanges() }
+        ApplicationManager.getApplication().invokeLaterOnWriteThread ({
+            ApplicationManager.getApplication().runWriteAction { sdkModificator.commitChanges() }
+        }, ModalityState.nonModal())
     }
 
     override fun suggestHomePath(): String? = suggestHomePaths().firstOrNull()

```

**Comment by joshuataylor** [comment]
Fantastic! Thanks, will take a look.

Legend 👏

**Comment by joshuataylor** [comment]
Yeah, it's an IntelliJ platform deprecation issue, so we need to go through and fix them. Another issue that arose is that 2025.1+ seems to break a whole bunch of weird and wonderful things with parsing - we might be able to hotfix some issues, but I'm blocked by a few IntelliJ Platform bugs. See https://github.com/KronicDeth/intellij-elixir/issues/3704

**Comment by joshuataylor** [comment]
@aikomastboom - This is great, thanks!

Similar to my original attempt with coroutines, I can see a bit of jankiness with the way IntelliJ handles JDKs/SDKs, especially in a background thread.

1. Add a Erlang SDK

2. Sometimes the classpaths/sourcepaths show, sometimes not (usually not).

<img width="1040" height="306" alt="Image" src="https://github.com/user-attachments/assets/b12eda0b-a480-4eee-bcc8-903a1a360e8e" />

3. Sometimes it works on clicking "Ok", then reopening.

Honestly not sure what the expected approach with this is, do we need our own JDK (SDK) loader or something?

**Comment by joshuataylor** [comment]
Cool! It's pretty much the same error - not a OS issue, a JetBrains deprecation issue :).

**Comment by joshuataylor** [comment]
This has also been an ongoing problem with PyCharm, where they have their
own weird SDK version as well. it's gotten worse with 2025.2, due to the
crackdown - so many core plugins are broken, crashing and returning issues
due to threading.

The workaround I see is adding a way to somehow add the SDK outside of
settings, as I believe refreshing settings in settings causes issues
(atlrast from my experience?)

**Comment by joshuataylor** [comment]
I'm going to pin this issue. I still think we need our own wizard or something.

**Comment by zubair-ali-iz** [comment]
https://github.com/KronicDeth/intellij-elixir/issues/3716

Is there anyone who can assist me in fixing this?
